recipes/gemma_full_finetune_distributed.py

```diff
@@ -0,0 +1,503 @@
+        # gradient_accumulation_steps param. This value is used for logging and tracking
+        # training state. The computation should happen after the dataloader has been setup
+        self._steps_per_epoch = (
+            len(self._dataloader) // self._gradient_accumulation_steps
+        )
+        if (
+            self.max_steps_per_epoch is not None
+            and self.max_steps_per_epoch < self._steps_per_epoch
+        ):
+            self._steps_per_epoch = self.max_steps_per_epoch
+        self.total_training_steps = self.epochs_run * self._steps_per_epoch
+
+    def _setup_model(
+        self,
+        cfg_model: DictConfig,
+        enable_activation_checkpointing: bool,
+        model_state_dict: Dict[str, Any],
+        shared: Dict[str, str] = None,
+    ) -> nn.Module:
+        """
+        Model initialization has some important considerations:
+            a. Initialize the model not on the meta device. This is because the model
+               is wrapped with FSDP which will shard the model across all available GPUs.
+            b. Shared weights are tied between the output layer and the token embeddings.
+               This is a temporary fix for GEMMA models.
+        """
+        init_start = time.perf_counter()
+
+        with utils.set_default_dtype(self._dtype):
+            model = config.instantiate(cfg_model)
+
+        log.info(
+            f"Model instantiation took {time.perf_counter() - init_start:.2f} secs"
+        )
+        # Load both the model weights
+        model.load_state_dict(model_state_dict)
+
+        # Tie the weights of the model
+        if shared is not None:  # Tie the weights of the model if required
+            for k, v in shared.weight_tying_config.items():
+                out = getattr(model, v)
+                tok = getattr(model, k)
+                out.weight = tok.weight
+                log.info(f"Model weights are shared between {k} and {v}.")
+
+        if self._dtype == torch.bfloat16:
+            model = model.to(torch.bfloat16)
+
+        # Wrap the model with FSDP. This will ensure that the model is sharded
+        # across all available GPUs.
+        model = FSDP(
+            module=model,
+            auto_wrap_policy=ModuleWrapPolicy({modules.TransformerDecoderLayer}),
+            sharding_strategy=torch.distributed.fsdp.ShardingStrategy.FULL_SHARD,
+            device_id=self._device,
+            # this recipe does not currently support mixed precision training
+            mixed_precision=None,
+            # Ensure we broadcast params and buffers from rank 0
+            sync_module_states=True,
+            # Initialize empty modules on all non-zero ranks
+            param_init_fn=(
+                lambda module: module.to_empty(
+                    device=torch.device("cuda"), recurse=False
+                )
+                if not self._is_rank_zero
+                else None
+            ),
+        )
```

Comment:
Sorry I missed this, but now that we are initializing on CPU across all ranks, we should set ```sync_module_states``` to False and ```param_init_fn``` to ```None```.

Reply:
Yeah, I update them in the latest version.